File: code/backend/src/main/java/com/taskflow/modules/notification/service/NotificationService.java
package com.taskflow.modules.notification.service;

import com.taskflow.common.PageResponse;
import com.taskflow.modules.notification.dto.CreateNotificationRequest;
import com.taskflow.modules.notification.dto.NotificationDto;
import com.taskflow.modules.notification.dto.UnreadCountDto;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Domain Service interface for managing in-app notifications, read status, and unread counts.
 */
public interface NotificationService {

    /**
     * Creates a new notification.
     *
     * @param request creation request payload
     * @return NotificationDto instance
     */
    NotificationDto createNotification(CreateNotificationRequest request);

    /**
     * Lists paginated notifications for a user.
     *
     * @param userId     UUID identifier of user
     * @param unreadOnly whether to filter only unread notifications
     * @param pageable   pagination parameters
     * @return PageResponse of NotificationDto items
     */
    PageResponse<NotificationDto> getUserNotifications(UUID userId, boolean unreadOnly, Pageable pageable);

    /**
     * Gets unread notification count for a user.
     *
     * @param userId UUID identifier of user
     * @return UnreadCountDto instance
     */
    UnreadCountDto getUnreadCount(UUID userId);

    /**
     * Marks a single notification as read.
     *
     * @param userId         UUID identifier of requesting user
     * @param notificationId UUID identifier of target notification
     * @return updated NotificationDto instance
     */
    NotificationDto markAsRead(UUID userId, UUID notificationId);

    /**
     * Marks all notifications as read for a user.
     *
     * @param userId UUID identifier of requesting user
     */
    void markAllAsRead(UUID userId);

    /**
     * Soft-deletes a notification.
     *
     * @param userId         UUID identifier of requesting user
     * @param notificationId UUID identifier of target notification
     */
    void deleteNotification(UUID userId, UUID notificationId);
}

File: code/frontend/src/features/notification/services/notification-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type {
  NotificationDto,
  PaginatedNotificationsResponse,
  UnreadCountResponse,
} from '../types';

export const notificationService = {
  getUserNotifications: async (
    page = 0,
    size = 20,
    unreadOnly = false
  ): Promise<PaginatedNotificationsResponse> => {
    const res = await apiClient.get<ApiResponse<PaginatedNotificationsResponse>>('/notifications', {
      params: { page, size, unreadOnly },
    });
    return res.data.data;
  },

  getUnreadCount: async (): Promise<UnreadCountResponse> => {
    const res = await apiClient.get<ApiResponse<UnreadCountResponse>>('/notifications/unread-count');
    return res.data.data;
  },

  markAsRead: async (notificationId: string): Promise<NotificationDto> => {
    const res = await apiClient.patch<ApiResponse<NotificationDto>>(`/notifications/${notificationId}/read`);
    return res.data.data;
  },

  markAllAsRead: async (): Promise<void> => {
    await apiClient.patch<ApiResponse<void>>('/notifications/read-all');
  },

  deleteNotification: async (notificationId: string): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`/notifications/${notificationId}`);
  },
};
